pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,206 @@
+		}
+		for _, instanceType := range instanceTypes.([]*ec2.InstanceTypeInfo) {
+			instanceTypeNames = append(instanceTypeNames, *instanceType.InstanceType)
+		}
+	}
+	return instanceTypeNames, nil
+}
+
+func (p *InstanceTypeProvider) getInstanceTypesForZone(ctx context.Context, zone string) ([]*ec2.InstanceTypeInfo, error) {
+	instanceTypes, err := p.getAllInstanceTypes(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving all instance types, %w", err)
+	}
+	instanceTypes, err = p.filterByZoneOfferings(ctx, instanceTypes, zone)
+	if err != nil {
+		return nil, fmt.Errorf("filtering instance types by zone offerings, %w", err)
+	}
+	return instanceTypes, nil
+}
+
+// getAllInstanceTypes retrieves all instance types from the ec2 DescribeInstanceTypes API using some opinionated filters
+func (p *InstanceTypeProvider) getAllInstanceTypes(ctx context.Context) ([]*ec2.InstanceTypeInfo, error) {
+	instanceTypes := []*ec2.InstanceTypeInfo{}
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("supported-virtualization-type"),
+				Values: []*string{aws.String("hvm")},
+			},
+		},
+	}
+	err := p.ec2.DescribeInstanceTypesPagesWithContext(ctx, describeInstanceTypesInput, func(page *ec2.DescribeInstanceTypesOutput, lastPage bool) bool {
+		instanceTypes = append(instanceTypes, page.InstanceTypes...)
+		return true
+	})
+	if err != nil {
+		return nil, fmt.Errorf("fetching instance types using ec2.DescribeInstanceTypes, %w", err)
+	}
+	zap.S().Debugf("Successfully discovered %d EC2 instance types", len(instanceTypes))
+	return instanceTypes, nil
+}
+
+// filterFrom returns a filtered list of instance types based on the provided resource constraints
+func (p *InstanceTypeProvider) filterFrom(instanceTypes []*ec2.InstanceTypeInfo, constraints *cloudprovider.Constraints) []*ec2.InstanceTypeInfo {
+	filtered := []*ec2.InstanceTypeInfo{}
+	architecture := utils.NormalizeArchitecture(*constraints.Architecture)
+
+	for _, instanceTypeInfo := range instanceTypes {
+		if (len(constraints.InstanceTypes) == 0 || functional.ContainsString(constraints.InstanceTypes, *instanceTypeInfo.InstanceType)) &&
+			(len(constraints.InstanceTypes) != 0 || p.isDefaultInstanceType(instanceTypeInfo)) &&
+			functional.ContainsString(aws.StringValueSlice(instanceTypeInfo.ProcessorInfo.SupportedArchitectures), architecture) &&
+			functional.ContainsString(aws.StringValueSlice(instanceTypeInfo.SupportedUsageClasses), "on-demand") {
+			filtered = append(filtered, instanceTypeInfo)
+		}
+	}
+	return filtered
+}
+
+// isDefaultInstanceType returns true if the instance type provided conforms to the default instance type criteria
+// This function is used to make sure we launch instance types that are suited for general workloads
+func (p *InstanceTypeProvider) isDefaultInstanceType(instanceTypeInfo *ec2.InstanceTypeInfo) bool {
+	if instanceTypeInfo.FpgaInfo == nil &&
```

Comment:
nit: could just return this (no `if` needed)

Reply:
whoops, copy/paste without thinking, 👍  done!